R/utilities-output-schema.R

```diff
@@ -77,3 +77,50 @@ clearOutputIntervals <- function(simulation) {
   simulation$outputSchema$clear()
   invisible(simulation)
 }
+
+#' Set end time of the simulation
+#' @description
+#' Either extends or shortens the simulation time to the specified end time.
+#' Time points that are later than the specified end time are removed.
+#' Intervals that start after the specified end time are removed.
+#' Intervals that start before the specified end time are shortened to the specified end time.
+#'
+#' @param simulation Simulation for which the end time should be set
+#' @param endTime End time of the simulation in min
+.setEndSimulationTime <- function(simulation, endTime) {
+  validateIsOfType(simulation, "Simulation")
+  validateIsNumeric(endTime)
+
+  # If the specified end time is outside of the current end time,
+  # extend the simulation time to the specified end time.
+  if (endTime > simulation$outputSchema$endTime) {
+    simulation$outputSchema$addTimePoints(timePoints = endTime)
+  } else {
+    # Otherwise, shorten the simulation time to the specified end time.
+    # First remove the time points that are later than the specified steady-state time.
+    timePoints <- simulation$outputSchema$timePoints
+    timePoints <- timePoints[timePoints <= endTime]
+
+    # Get the old output intervals
+    oldOutputIntervals <- simulation$outputSchema$intervals
+
+    # Clear the output schema
+    simulation$outputSchema$clear()
+    # Iterate through output intervals and only add those that start before the
+    # specified steady-state time.
+    for (outputInterval in oldOutputIntervals) {
+      if (outputInterval$startTime$value <= endTime) {
+        addOutputInterval(
+          simulation = simulation,
+          startTime = outputInterval$startTime$value,
+          endTime = min(outputInterval$endTime$value, endTime),
+          resolution = outputInterval$resolution$value,
+          intervalName = outputInterval$name
+        )
+      }
+    }
+
+    # Don't forget to add latest time
+    simulation$outputSchema$addTimePoints(c(timePoints, endTime))
+  }
```

Comment:
Here you could add a check that either at least 1 output interval or at least 1 (2?) single output time points are in the final output schema (otherwise the simulation will not produce any results).

Also what would happen if the simulation has e.g. the only output interval [10h .. 20h] and I set the end time to 5h?

Reply:
> Here you could add a check that either at least 1 output interval or at least 1 (2?) single output time points are in the final output schema (otherwise the simulation will not produce any results).

I think this is not required because `endTime` will always be a valid numeric value, and it is always added to the time schema. But I added a validation that it is >0.

> Also what would happen if the simulation has e.g. the only output interval [10h .. 20h] and I set the end time to 5h?

The output interval is removed, only one point 5h is retained. Added a test for this.